lib/WUI/nhttp/gcode_upload.cpp

```diff
@@ -222,22 +222,67 @@ UploadHooks::Result GcodeUpload::data(std::string_view data) {
 
 namespace {
 
-    UploadHooks::Result prepend_usb_path(const char *filename, char *filepath_out, size_t filepath_size) {
-        const uint32_t fname_length = strlen(filename);
-
-        if ((fname_length + USB_MOUNT_POINT_LENGTH) >= filepath_size) {
+    UploadHooks::Result prepend_usb_path(string_view filename, char *filepath_out, size_t filepath_size) {
+        if ((filename.length() + USB_MOUNT_POINT_LENGTH) >= filepath_size) {
             // The Request header fields too large is a bit of a stretch...
             return make_tuple(Status::RequestHeaderFieldsTooLarge, "File name too long");
         }
         strlcpy(filepath_out, USB_MOUNT_POINT, USB_MOUNT_POINT_LENGTH + 1);
-        strlcat(filepath_out, filename, FILE_PATH_BUFFER_LEN - USB_MOUNT_POINT_LENGTH);
+        strncat(filepath_out, filename.data(), filename.length());
+        filepath_out[USB_MOUNT_POINT_LENGTH + filename.length()] = '\0';
+
+        return make_tuple(Status::Ok, nullptr);
+    }
+
+    UploadHooks::Result make_dir(string_view dir) {
+        char path[USB_MOUNT_POINT_LENGTH + dir.length() + 1];
+
+        auto error = prepend_usb_path(dir, path, sizeof(path));
+        if (std::get<0>(error) != Status::Ok)
+            return error;
+        if (mkdir(path, 777) != 0) {
+            // teoretically could fail not on the first
+            // folder and we would then leave some previous
+            // folder in place ,but this is improbable...
+            if (errno != EEXIST)
+                return make_tuple(Status::InternalServerError, "Failed to create a folder");
+        }
+        return make_tuple(Status::Ok, nullptr);
+    }
+
+    UploadHooks::Result make_dirs(string_view path) {
+        size_t pos = path.find('/');
+        while (pos != path.npos) {
+            if (auto error = make_dir(path.substr(0, pos)); std::get<0>(error) != Status::Ok)
+                return error;
+            pos = path.find('/', pos + 1);
+        }
 
         return make_tuple(Status::Ok, nullptr);
     }
 
+    //Note: if we would have weird name chars, it would fail
+    // only at the end of upload, but if we would create all the
+    // dirs to check, then we would have to also remove them
+    // otherwise we might create leftovers if the upload failed
+    // for whatever reason, so for simplicity and performance
+    // this is probably fine
+    std::variant<bool, UploadHooks::Result> check_is_needed(string_view path) {
+        size_t pos = path.find_last_of('/');
+        if (pos == path.npos)
+            return true;
+        char last_dir[pos + USB_MOUNT_POINT_LENGTH + 1];
+        auto error = prepend_usb_path(path.substr(0, pos), last_dir, sizeof(last_dir));
+        if (std::get<0>(error) != Status::Ok)
+            return error;
```

Comment:
Don't we have some function somewhere for this? `basename` or so? :thinking:

Reply:
`basename` returns the name of the file, meaning everything _after_ the last `/`, I want the parent path (everything _before_ `/`), I have not been able to find any function, that would do that, but I might have missed it somewhere...